src/System.Text.Primitives/System/Text/Encoders/Utf8Util.Helpers.cs

```diff
@@ -0,0 +1,441 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+//
+// This file contains utility methods shared by the UTF-8 workhorse methods.
+//
+
+namespace System.Buffers.Text
+{
+    internal static partial class Utf8Util
+    {
+        /// <summary>
+        /// Casts an <see cref="IntPtr"/> to an <see cref="int"/> without overflow checking.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int ConvertIntPtrToInt32WithoutOverflowCheck(IntPtr value)
+        {
+            if (IntPtr.Size == 4)
+            {
+                return (int)value;
+            }
+            else
+            {
+                return (int)(long)value;
+            }
+        }
+
+        /// <summary>
+        /// Returns <see langword="true"/> iff all bytes in <paramref name="value"/> are ASCII.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool DWordAllBytesAreAscii(uint value)
+        {
+            return ((value & 0x80808080U) == 0U);
+        }
+
+        /// <summary>
+        /// Given a UTF-8 buffer which has been read into a DWORD in machine endianness,
+        /// returns <see langword="true"/> iff the buffer contains two UTF-8 sequences
+        /// that match the mask [ 110yyyyy 10xxxxxx 110yyyyy 10xxxxxx ]. This method *does not*
+        /// validate that the sequences are well-formed; the caller must still perform
+        /// overlong form checking.
```

Comment:
What do you mean by "overlong form checking"?

Reply:
"Overlong" is the Unicode term for a UTF-8 sequence that consists of more than the minimum number of code units required to represent the scalar value. For instance, it is forbidden to use a 2-byte sequence to represent an ASCII character.